File: CODE/leetcodes/src/main/java/com/hit/math/array_string/_14.java
package com.hit.math.array_string;

/**
 * 14. Longest Common Prefix
 * <p>
 * Write a function to find the longest common prefix string amongst an array of strings.
 * <p>
 * If there is no common prefix, return an empty string "".
 * <p>
 * Example 1:
 * <p>
 * Input: ["flower","flow","flight"]
 * Output: "fl"
 * <p>
 * Example 2:
 * <p>
 * Input: ["dog","racecar","car"]
 * Output: ""
 * <p>
 * Explanation: There is no common prefix among the input strings.
 * <p>
 * Note:
 * <p>
 * All given inputs are in lowercase letters a-z.
 */
public class _14 {
    public static String longestCommonPrefix(String[] strs) {
        if (strs == null || strs.length == 0) {
            return "";
        }
        /**
         * Let first element of strs as a initial aim longest common prefix
         */
        String aimPrefix = strs[0];
        int i = 1;
        /**
         * Circle strs.length - 1 times
         */
        while (i < strs.length) {
            /**
             * This strs[i].indexOf(aimPrefix) != 0 condition means
             * both param have't same common prefix
             */
            while (strs[i].indexOf(aimPrefix) != 0) {
                /**
                 * Because common prefix is must be satisfied all element in strs,
                 * so, any two elements can decide this longest common prefix.
                 *
                 * If strs[i].indexOf(aimPrefix) != 0 is true,
                 * we can truncate aimPrefix many a time
                 */
                aimPrefix = aimPrefix.substring(0, aimPrefix.length() - 1);
            }

            /**
             * If the first cycle is completed, aimPrefix.length() == 0 is true
             * it's mean there is no longest prefix surely
             */
            if (aimPrefix.length() == 0) {
                return "";
            } else {
                /**
                 * i++ means this circle handled and begin next circle judgement
                 */
                i++;
            }
        }
        return aimPrefix;
    }

    public static void main(String[] args) {
        String[] strs = {"follow", "a", "fob"};
        String[] strs2 = {"follow", "fo", "fol"};
        System.out.println(_14.longestCommonPrefix(strs));
        System.out.println(_14.longestCommonPrefix(strs2));
    }
}

File: codes/go/leetcodes/interview/easy_collection/14.go
package easy_collection

import "strings"

/**
 * 14. Longest Common Prefix
 * <p>
 * Write a function to find the longest common prefix string amongst an array of strings.
 * <p>
 * If there is no common prefix, return an empty string "".
 * <p>
 * Example 1:
 * <p>
 * Input: ["flower","flow","flight"]
 * Output: "fl"
 * <p>
 * Example 2:
 * <p>
 * Input: ["dog","racecar","car"]
 * Output: ""
 * <p>
 * Explanation: There is no common prefix among the input strings.
 * <p>
 * Note:
 * <p>
 * All given inputs are in lowercase letters a-z.
 */

func longestCommonPrefix(strs []string) string {
	if len(strs) == 0 {
		return ""
	}
	aimStr := strs[0]
	index := 1
	for index < len(strs) {
		for !strings.HasPrefix(strs[index], aimStr) {
			aimStr = aimStr[0 : len(aimStr)-1]
		}
		if len(aimStr) == 0 {
			return ""
		}
		index++
	}
	return aimStr
}
